### PR TITLE
Fix CASE statement when branch expression is of String Datatype.

### PR DESCRIPTION
**Pull request by Yvinayak07**
### 1. Issue:

Return type of overall CASE expression when branch expression is of String Datatype, is evaluated as per Postgres behaviour, which gives a different Return type as compared to TSQL. 

eg. Return type of overall CASE expression is evaluated to become VARCHAR, if all branches are of NVARCHAR type because Postgres tries to coerce result of every branch expression to base type (VARCHAR), whereas in TSQL, overall return type is NVARCHAR in this example.

```
CREATE TABLE ALLTEXTS(
CODE nvarchar(30) NOT NULL,
LANG nvarchar(2) NOT NULL,
TEXT nvarchar(2000) NULL
)

INSERT INTO ALLTEXTS (CODE, LANG, TEXT)
VALUES ('53075', 'EN', N'Registration Card - Standard');

INSERT INTO ALLTEXTS (CODE, LANG, TEXT)
VALUES ('53075', 'AR', N'بطاقة التسجيل - قياسية');

INSERT INTO ALLTEXTS (CODE, LANG, TEXT)
VALUES ('53075', 'JP', N'登録カード–標準');

SELECT CASE WHEN 1 = 1 THEN TEXT END AS WrongValue2 FROM ALLTEXTS
```
Return type: varchar
expected type: nvarchar

### 2. Changes made to fix the issues

tsql_select_common_type_hook() is set to return common_type for all branches when branch expression is of String Datatype, that calculate according to the TSQL precedence, whereas till now it was set according to the PG precedence order (Which was different from TSQL).This hook solved our problem as it returns result according to the TSQL precedence order. We have also fixed the typmod issue, by calculating the common typmod and setting all branches of CASE statement to the calculated typmod.

Task: BABEL-5103, BABEL-4332
Authored-by: yashneet vinayak [yashneet@amazon.com](mailto:yashneet@amazon.com)
Signed-off-by: yashneet vinayak [yashneet@amazon.com](mailto:yashneet@amazon.com)

### Issues Resolved
Task: BABEL-5103, BABEL-4332

Engine Side PR link: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/452

### Test Scenarios Covered ###
* **Use case based -**Yes

* **Boundary conditions -**

* **Arbitrary inputs -**Yes

* **Negative test cases -**

* **Minor version upgrade tests -**Yes

* **Major version upgrade tests -**Yes

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**Yes

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 1]
```diff
@@ -1776,7 +1778,8 @@ tsql_select_common_type_hook(ParseState *pstate, List *exprs, const char *contex
 			strncmp(context, "INTERSECT", strlen("INTERSECT")) == 0 ||
 			strncmp(context, "EXCEPT", strlen("EXCEPT")) == 0 ||
 			strncmp(context, "VALUES", strlen("VALUES")) == 0 ||
-			strncmp(context, "UNION/INTERSECT/EXCEPT", strlen("UNION/INTERSECT/EXCEPT")) == 0)
+			strncmp(context, "UNION/INTERSECT/EXCEPT", strlen("UNION/INTERSECT/EXCEPT")) == 0 ||
+			strncmp(context, "CASE", strlen("CASE")) == 0)
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Suggestion: can we rewrite use of strncmp [like this](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/e57f03a5585b21043b6b26ab23894068e9250fbb/src/backend/nodes/readfuncs.c#L361)
Get length of context once and then use it to filter out most conditions. Would save some time.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 2]
```diff
@@ -1487,7 +1487,9 @@ is_tsql_char_type_with_len(Oid type)
 	return utilptr->is_tsql_bpchar_datatype(type) ||
 			utilptr->is_tsql_nchar_datatype(type) ||
 			utilptr->is_tsql_varchar_datatype(type) ||
-			utilptr->is_tsql_nvarchar_datatype(type);
+			utilptr->is_tsql_nvarchar_datatype(type) ||
+			utilptr->is_tsql_text_datatype(type) ||
+			utilptr->is_tsql_ntext_datatype(type);
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Why are we adding [N]TEXT here and not letting PG to decide the common type for these types? I assume that, PG will have different precedence and we don't want PG to decide that. Lets add proper comment in code to explain that.

Were we observing the similar issues for UNION as well? Lets add few testcases for UNION to test that.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 1 above)
Comment (sumitj824) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: @Yvinayak07 Please address this.

File: test/JDBC/expected/BABEL-5103-vu-prepare.out [hunk 3]
```diff
@@ -0,0 +1,54 @@
+CREATE TABLE BABEL_5103_T1(
```
Comment (sumitj824) [inline test/JDBC/expected/BABEL-5103-vu-prepare.out]: Lets add meaningful name of this testfile. And add testcase for NULL and local variables if not added already.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 1 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Fixed this change

File: test/JDBC/expected/BABEL-5103-vu-prepare.out (hunk 3 above)
Comment (Yvinayak07) [reply test/JDBC/expected/BABEL-5103-vu-prepare.out]: Renamed file to BABEL-CASE_EXPR. 
Also added Null tests

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 2 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Added these conditions for CASE only (separate function), So no need to add union tests.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 4]
```diff
@@ -1763,21 +1775,23 @@ static Oid
...
+	if (sql_dialect != SQL_DIALECT_TSQL || !context)
 		return InvalidOid;
-
-	if (!context)
-		return InvalidOid;
-	else if (strncmp(context, "ISNULL", strlen("ISNULL")) == 0)
+    
+	length = strlen(context);
+	
+	if (strncmp(context, "ISNULL", strlen("ISNULL")) == 0)
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Please use similar checks as done below like `(len == 6 && strncmp(context, "ISNULL", 6) == 0)` for all occurences in this function.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 5]
```diff
@@ -1827,15 +1841,17 @@ tsql_handle_constant_literals_hook(ParseState *pstate, Node *e)
  * output type based on TSQL's precedence rules
  */ 
 static Oid
-select_common_type_setop(ParseState *pstate, List *exprs, Node **which_expr)
+select_common_type_setop(ParseState *pstate, List *exprs, Node **which_expr, const char *context)
 {
 	Node		*result_expr = (Node*) linitial(exprs);
 	Oid			result_type = InvalidOid;
 	ListCell	*lc;
+    bool        isCase = (strncmp(context, "CASE", 4) == 0) ? true : false;
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Instead of `(strncmp(context, "CASE", 4) == 0)`, it should be `(strlen(context) == 4 && strncmp(context, "CASE", 4) == 0)`

NIT: Please maintain consistent naming convention for variable names and indentation within the function.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 6]
```diff
@@ -1845,7 +1861,8 @@ select_common_type_setop(ParseState *pstate, List *exprs, Node **which_expr)
 			continue;
 		else if (is_tsql_str_const(expr))
 			type = common_utility_plugin_ptr->lookup_tsql_datatype_oid("varchar");
-		else if (!is_tsql_char_type_with_len(type))
+		else if ((!isCase && !is_tsql_char_type_with_len(type)) ||
+		         (isCase && !is_tsql_char_type_with_len_case_exp(type)))
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Instead of creating new function `is_tsql_char_type_with_len_case_exp` and invoking it separately, I would suggest to extend the definition of `is_tsql_char_type_with_len` by passing extra arg so that we don't make multiple calls on functions like `is_tsql_nvarchar_datatype`.

Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: is_tsql_char_type_with_len   if we change the function defination for this we need to change it at other place(used in this function also tsql_select_common_typmod_hook) also Where we don't have context.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 5 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: chnaged the variable name to is_case_expr and added strlen(context) == 4

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 4 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Addressed this change.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 6 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Made the chnage just kept one function to reduce the redundant code.

File: test/JDBC/expected/babel_choose.out [hunk 7]
```diff
@@ -84,7 +84,7 @@ GO
 select choose (choose (1, 2, 3), 'a', 'b', 'c');
 GO
 ~~START~~
-text
+varchar
 b
 ~~END~~
```
Comment (rohit01010) [inline test/JDBC/expected/babel_choose.out:89]: Why behaviour of choose got changed, does choose internally uses `CASE EXPRESSION` ?

File: test/JDBC/expected/BABEL-1524.out [hunk 8]
```diff
@@ -80,7 +80,7 @@ go
 select IIF(timestamp = NULL, 'null', 'not null') from t1;
 go
 ~~START~~
-text
+varchar
 not null
 not null
 ~~END~~
```
Comment (rohit01010) [inline test/JDBC/expected/BABEL-1524.out:86]: why behaviour of `IFF` function got changed ? Does IFF uses `CASE EXPRESSION` internally ?

File: test/JDBC/expected/database_metadata.out [hunk 9]
```diff
@@ -47,14 +47,14 @@ master#!#dbo#!#dbmeta_view1#!#VIEW#!#<NULL>
 
 dbmeta#!#getColumns#!#master|dbo|dbmeta_tab1
 ~~START~~
-varchar#!#varchar#!#varchar#!#varchar#!#smallint#!#varchar#!#int#!#int#!#smallint#!#smallint#!#smallint#!#varchar#!#nvarchar#!#smallint#!#smallint#!#int#!#int#!#varchar#!#int#!#int#!#int#!#tinyint#!#text#!#text#!#smallint#!#smallint#!#nvarchar#!#nvarchar#!#nvarchar#!#nvarchar#!#nvarchar#!#nvarchar
+varchar#!#varchar#!#varchar#!#varchar#!#smallint#!#varchar#!#int#!#int#!#smallint#!#smallint#!#smallint#!#varchar#!#nvarchar#!#smallint#!#smallint#!#int#!#int#!#varchar#!#int#!#int#!#int#!#tinyint#!#varchar#!#varchar#!#smallint#!#smallint#!#nvarchar#!#nvarchar#!#nvarchar#!#nvarchar#!#nvarchar#!#nvarchar
```
Comment (rohit01010) [inline test/JDBC/expected/database_metadata.out:50]: how is the fix for case expression result type is affecting here ?

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 10]
```diff
@@ -1481,13 +1482,23 @@ tsql_func_select_candidate(List *names,
...
-			utilptr->is_tsql_varchar_datatype(type) ||
-			utilptr->is_tsql_nvarchar_datatype(type);
+	result =  utilptr->is_tsql_bpchar_datatype(type) ||
+			  utilptr->is_tsql_nchar_datatype(type) ||
+			  utilptr->is_tsql_varchar_datatype(type) ||
+			  utilptr->is_tsql_nvarchar_datatype(type);
+	
+	/* For case expr we need to find common type based on TSQL's
+	 * precedence for text and ntext also.
+	 */
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: NIT: Comment should like
```
/*
 * This is comment.
 */
```

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 11]
```diff
@@ -1481,13 +1482,23 @@ tsql_func_select_candidate(List *names,
 }
 
 static bool
-is_tsql_char_type_with_len(Oid type)
+is_tsql_char_type_with_len(Oid type, bool is_case_expr)
 {
+	bool					result = false;
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: NIT: No need of setting this explicitly to false.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 12]
```diff
@@ -2003,6 +2019,58 @@ tsql_select_common_typmod_hook(ParseState *pstate, List *exprs, Oid common_type)
 	return max_typmods;
 }
 
+/* For CASE expression, this function will calculate typmod from tsql_select_common_typmod_hook(),
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: NIT: comment style.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 13]
```diff
@@ -2003,6 +2019,58 @@ tsql_select_common_typmod_hook(ParseState *pstate, List *exprs, Oid common_type)
...
+ * and after that we will set this typmod to all the CASE branches from coerce_to_target_type().
+ */
+static Node*
+tsql_set_common_typemod_case_expr_hook(ParseState *pstate, List *exprs, CaseExpr *newc)
+{
+	CaseExpr   *tmpc  =  makeNode(CaseExpr);
+	int32       typmod;
+        ListCell   *l;
+
+        if(sql_dialect != SQL_DIALECT_TSQL)
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Shouldn't this dialect check be part of engine code itself?

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 14]
```diff
@@ -2003,6 +2019,58 @@ tsql_select_common_typmod_hook(ParseState *pstate, List *exprs, Oid common_type)
...
+                return (Node *) newc;
+
+        tmpc->xpr = newc->xpr;
+        tmpc->casecollid = newc->casecollid;
+        tmpc->location = newc->location;
+        tmpc->arg = newc->arg;
+        tmpc->args = newc->args;
+        tmpc->defresult = newc->defresult;
+        tmpc->casetype = newc->casetype;
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Do we really need new copy? Can't we directly update the previous one?

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 15]
```diff
@@ -2003,6 +2019,58 @@ tsql_select_common_typmod_hook(ParseState *pstate, List *exprs, Oid common_type)
...
+
+        // calculating common_typemod for case exper
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: NIT: comment style.

File: test/JDBC/expected/BABEL-1524.out (hunk 8 above)
Comment (Yvinayak07) [reply test/JDBC/expected/BABEL-1524.out:86]: Yes IFF uses CASE implemetation.

File: test/JDBC/expected/babel_choose.out (hunk 7 above)
Comment (Yvinayak07) [reply test/JDBC/expected/babel_choose.out:89]: Yes choose uses CASE implemetation.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 15 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Resolved

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 14 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: resolved

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 13 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: resolved

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 12 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: resolved

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 11 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: resolved

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 10 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: resolved

File: test/JDBC/expected/database_metadata.out (hunk 9 above)
Comment (Yvinayak07) [reply test/JDBC/expected/database_metadata.out:50]: After running this JDBC tests, I have generated the queries used in "dbmeta#!#getColumns#!#master|dbo|dbmeta_tab1" inside logfile. It is seen that it is implemented using CASE statement. 
logfile:
"
sp_executesql: handle: -1 statement: select [master].sys.schemas.name 'TABLE_SCHEM', CASE WHEN [master].sys.schemas.name  IN  ('dbo', 'guest','INFORMATION_SCHEMA','sys','db_owner', 'db_accessadmin', 'db_securityadmin', 'db_ddladmin'  ,'db_backupoperator','db_datareader','db_datawriter','db_denydatareader','db_denydatawriter')  THEN null ELSE 'master'  END 'TABLE_CATALOG'    from [master].sys.schemas where [master].sys.schemas.name like @P0   order by 2, 1"
"
sql_batch statement: select [master].sys.schemas.name 'TABLE_SCHEM', CASE WHEN [master].sys.schemas.name  IN  ('dbo', 'guest','INFORMATION_SCHEMA','sys','db_owner', 'db_accessadmin', 'db_securityadmin', 'db_ddladmin'  ,'db_backupoperator','db_datareader','db_datawriter','db_denydatareader','db_denydatawriter')  THEN null ELSE 'master'  END 'TABLE_CATALOG'    from [master].sys.schemas order by 2, 1"

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 16]
```diff
@@ -2003,6 +2020,47 @@ tsql_select_common_typmod_hook(ParseState *pstate, List *exprs, Oid common_type)
...
+
+        newc->defresult = (Expr *) 
+                coerce_to_target_type(pstate,
+                                (Node *) newc->defresult, 
+                                newc->casetype, 
+                                newc->casetype, 
+                                typmod, 
+                                COERCION_ASSIGNMENT,
+                                COERCE_IMPLICIT_CAST,
+                                -1);
```
Comment (rohit01010) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Here we only want to set typmod of the CASE branches, why are we not using `coerce_type_typmod` function instead of `coerce_to_target_type` ?

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 17]
```diff
@@ -2003,6 +2020,47 @@ tsql_select_common_typmod_hook(ParseState *pstate, List *exprs, Oid common_type)
...
+        /* calculating common_typemod for case expr */
+        typmod = tsql_select_common_typmod_hook(pstate, exprs, newc->casetype);
+
+        newc->defresult = (Expr *) 
+                coerce_to_target_type(pstate,
+                                (Node *) newc->defresult, 
+                                newc->casetype, 
+                                newc->casetype, 
+                                typmod, 
+                                COERCION_ASSIGNMENT,
```
Comment (rohit01010) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c]: why are we using `COERCION_ASSIGNMENT ` and not `COERCION_IMPLICIT` ?

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 16 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: coerce_type_typmod is a static function so it couldn't be invoked in some other file.

Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: Removed coerce_type_typmod function as static and used this in place of coerce_to_target_type.

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 17 above)
Comment (Yvinayak07) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c]: COERCION_ASSIGNMENT is for user defined datatype, for internal use we use COERCION_IMPLICIT. So changing it to COERCION_IMPLICIT.

**Review by rohit01010 — APPROVED** [review]
LGTM

**Review by sumitj824 — APPROVED** [review]

**Review by Deepesh125 — DISMISSED** [review]

File: test/JDBC/expected/non_default_server_collation/chinese_prc_ci_as/BABEL-CASE_EXPR-vu-verify.out [hunk 18]
```diff
@@ -0,0 +1,5368 @@
...
+END AS RESULT
+GO
+~~START~~
+nvarchar
+بطاقة التسجيل - قياسية
+~~END~~
+
+
+SELECT CASE N'  abc🙂defghi🙂🙂    ' COLLATE ARABIC_CI_AS
+   WHEN 'abcd' COLLATE japanese_ci_as THEN 'abcd'
```
Comment (KushaalShroff) [inline test/JDBC/expected/non_default_server_collation/chinese_prc_ci_as/BABEL-CASE_EXPR-vu-verify.out:5358]: LGTM but we had an ask to put COLLATE as the resultant node as well

Comment (Yvinayak07) [reply test/JDBC/expected/non_default_server_collation/chinese_prc_ci_as/BABEL-CASE_EXPR-vu-verify.out:5358]: Added those tests.

**Review by KushaalShroff — APPROVED** [review]
